### PR TITLE
Port over to AutoML common

### PR DESCRIPTION
**Pull request by eddiebergman**
* Adds the `automl_common` submodule and uses it's backend for proof of concept
* Updates the `MANIFEST.in` file to not include the test files of `automl_common`
* Adds section to installation about installing from source (now that it's not as straight forward with submodules)
* Updates workflow files to install submodules correctly

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "automl_common"]
+	path = automl_common
```
[the remark this answers is not in the document]
Comment (eddiebergman) [reply .gitmodules:2]: Okay, I see your point but I would also say that anyone who is installing with `pip` will never see any of this. Only github clone and install'ers will see it, in which case transparency that there is a subpackage might be nice.

On a more practical note, it also means the tests distributed with `automl_common` don't end up in the source code of autosklearn.

File: README.md [hunk 2]
```diff
@@ -22,6 +22,27 @@ cls.fit(X_train, y_train)
 predictions = cls.predict(X_test)
 ```
 
+## Install from source
```
[the remark this answers is not in the document]
Comment (eddiebergman) [reply README.md:25]: Fair enough, my thoughts were that `git` installers will come from github and hence see the github README where they need to be made aware of `--recurse-submodules`. I will add it to the docs and remove it here anyways, it makes everything in one place which is easier to maintain.

File: .gitmodules (hunk 1 above)
[the remark this answers is not in the document]
Comment (mfeurer) [reply .gitmodules:2]: Hey, I just had another look and I think there is another issue here: the `.tar.gz` doesn't include the submodule as it only packages everything under `autosklearn`. So I think that `automl_common` actually has to live under `autosklearn`. Test etc can be excluded with `exclude` in the [`MANIFEST.in`](https://packaging.python.org/guides/using-manifest-in/).

However, this raises the important issue that the CI wasn't strict enough. We need to go to a different directory to run the tests as [done in scikit-learn](https://github.com/scikit-learn/scikit-learn/blob/main/build_tools/github/test_source.sh#L17) to make sure we don't accidentally import what's in the current working directory.

[the remark this answers is not in the document]
Comment (eddiebergman) [reply .gitmodules:2]: It can definitely live outside autosklearn and just build the package differently but sure, I don't mind either way,  we can exclude them that way.

* I will put `automl_common` under autosklearn.
* I will see how to test that `MANIFEST.in` works with excluding test and other files.
* I will take a look at the sklearn things

File: MANIFEST.in [hunk 3]
```diff
@@ -1,9 +1,16 @@
...
-recursive-include autosklearn/experimental/ *.json
-include autosklearn/experimental/askl2_training_data.json
-include LICENSE.txt
+
+# Meta-data
+recursive-include autosklearn/metalearning/files *.arff *.csv *.txt
+recursive-include autosklearn/experimental *.json
+
+# Remove tests and workflow files from automl_common
+prune autosklearn/automl_common/test
```
Comment (mfeurer) [inline MANIFEST.in:11]:
```suggestion
prune autosklearn/automl_common/test
exclude autosklearn/automl_common/setup.py
```

File: README.md [hunk 4]
```diff
@@ -1,4 +1,4 @@
-# auto-sklearn
+auto-sklearn
```
Comment (mfeurer) [inline README.md:1]: What's the reason to drop the `#`?

File: test/test_util/test_backend.py [hunk 5]
```diff
@@ -3,7 +3,7 @@
 import unittest
 import unittest.mock
 
-from autosklearn.util.backend import Backend
+from autosklearn.automl_common.common.utils.backend import Backend
```
Comment (mfeurer) [inline test/test_util/test_backend.py:6]: Shouldn't these tests live in `automl_common` now?

File: .gitmodules (hunk 1 above)
[the remark this answers is not in the document]
Comment (eddiebergman) [reply .gitmodules]: * I put it underneath and it seems to work
* I updated the manifest to not include automl_commons tests
* Regarding testing, our test workflow already specifies to use the correct autosklearn test folder

File: scripts/2015_nips_paper/run/score_ensemble.py [hunk 6]
```diff
@@ -102,9 +102,11 @@ def main(input_directories, output_file, task_id, seed, ensemble_size, n_jobs=1)
     losses = []
     top_models_at_step = dict()
 
-    backend = create(input_directory, input_directory + "_output",
-                     delete_tmp_folder_after_terminate=False,
-                     shared_mode=True)
+    backend = create(
```
Comment (mfeurer) [inline scripts/2015_nips_paper/run/score_ensemble.py:105]: Sorry, might be a stupid question, but don't we require a prefix here, too?

File: test/test_evaluation/test_abstract_evaluator.py [hunk 7]
```diff
@@ -294,11 +298,14 @@ def test_add_additional_components(self):
 
         context = BackendContext(
             temporary_directory=os.path.join(self.working_directory, 'tmp'),
+            output_directory=os.path.join(self.working_directory, 'tmp_output'),
```
Comment (mfeurer) [inline test/test_evaluation/test_abstract_evaluator.py:301]: Do we need to pass in the output directory? Auto-sklearn doesn't use this concept any more.

File: doc/installation.rst [hunk 8]
```diff
@@ -100,6 +100,30 @@ to read in more details check
...
+
+    # Install it in editable mode with all optional dependencies
+    pip install -e ".[test,doc,examples]"
+
+We use submodules so you will have to make sure the submodule is initialized if you
+missed the `--recurse-submodules` option.
+
+.. code:: bash
+
+    git clone git@github.com:automl/auto-sklearn.git
```
Comment (mfeurer) [inline doc/installation.rst:121]: Should we mention this in the contribution guide, too?

File: .gitmodules (hunk 1 above)
[the remark this answers is not in the document]
Comment (mfeurer) [reply .gitmodules]: > Regarding testing, our test workflow already specifies to use the correct autosklearn test folder

Do we `cd` in some other directory? I can't find that right now. What I meant is that we execute the tests from for example `/tmp/tests` so that we don't by accident import anything from the current working directory.

[the remark this answers is not in the document]
Comment (eddiebergman) [reply .gitmodules]: I don't see why that should now be an issue? We'll see what the tests say

File: MANIFEST.in (hunk 3 above)
Comment (eddiebergman) [reply MANIFEST.in:11]: Good catch, i"ll update this

File: test/test_util/test_backend.py (hunk 5 above)
Comment (eddiebergman) [reply test/test_util/test_backend.py:6]: They should, however the new backend redoes the tests so I'll change them then.

File: scripts/2015_nips_paper/run/score_ensemble.py (hunk 6 above)
Comment (eddiebergman) [reply scripts/2015_nips_paper/run/score_ensemble.py:105]: We do, this script is broken I'm fairly sure but I'll fix this

File: test/test_evaluation/test_abstract_evaluator.py (hunk 7 above)
Comment (eddiebergman) [reply test/test_evaluation/test_abstract_evaluator.py:301]: It's unfortunately a required argument, AutoPytorch doesn't use it either so it'll be fixed with the new backend of automl_common

File: doc/installation.rst (hunk 8 above)
Comment (eddiebergman) [reply doc/installation.rst:121]: Definitely, I'll add this

File: .gitmodules (hunk 1 above)
[the remark this answers is not in the document]
Comment (eddiebergman) [reply .gitmodules]: tests seem to run without problem and it only runs `autosklearn` tests

File: test/test_evaluation/test_abstract_evaluator.py (hunk 7 above)
Comment (eddiebergman) [reply test/test_evaluation/test_abstract_evaluator.py:301]: Funnily enough, it's typed as `Optional[str]` but doesn't provide a default parameter of `None` s it's required. It's updated in the new backend so I will leave it for now.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -30,16 +30,20 @@ Following that we'll tell you about how you can test your changes locally and th
...
-    git clone git@github.com:your-username/auto-sklearn.git
+    git clone --recurse-submodules git@github.com:your-username/auto-sklearn.git
 
     # Navigate into the cloned repo
     cd auto-sklearn
 
     # Create a new branch based off the development one
     git checkout -b my_new_branch development
+    
+    git submodule udate --init --recursive  # If you missed the --recurse-submodules arg
```
Comment (mfeurer) [inline CONTRIBUTING.md:46]: What do you think about? This would prevent users from accidentally executing that line (not sure if it would do any harm, though).
```suggestion
    # Execute the following line if you missed the --recurse-submodules arg
    # git submodule udate --init --recursive
```

Comment (eddiebergman) [reply CONTRIBUTING.md:46]: It shouldn't but make's sense

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -50,6 +56,8 @@ Following that we'll tell you about how you can test your changes locally and th
 
     # Create a new branch based on the currently active branch
     git checkout -b my_new_branch
+
+    git submodule udate --init --recursive  # If you missed the --recurse-submodules arg
```
Comment (mfeurer) [inline CONTRIBUTING.md:60]:
```suggestion
    # If you missed the --recurse-submodules arg during clone or need to install the
    # submodule manually, then execute the following line:
    # git submodule udate --init --recursive
```

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -335,6 +343,9 @@ Lastly, if the feature really is a game changer or you're very proud of it, cons
     cd auto-sklearn
     git checkout -b my_new_branch development
 
+    # Initaize autosklearn/automl_common submodule
```
Comment (mfeurer) [inline CONTRIBUTING.md:346]:
```suggestion
    # Initialize autosklearn/automl_common submodule
```

**Review by mfeurer — APPROVED** [review]
Two more typos, then we're good to go :)

File: CONTRIBUTING.md (hunk 11 above)
Comment (eddiebergman) [reply CONTRIBUTING.md:346]: I should honestly consider having a spell checker on when looking at `.md` files and `.rst` files ...

Comment (mfeurer) [reply CONTRIBUTING.md]: https://github.com/marketplace/actions/github-spellcheck ;)

**Review by mfeurer — APPROVED** [review]
